lib/main.js

```diff
@@ -205,4 +219,120 @@ module.exports = class Client {
+                writer.transform(entry.transform, entry.options);
+            });
+
+            this.plugins.forEach(entry => {
+                writer.plugin(entry.plugin, entry.options);
+            });
+
+            return writer;
+        }
+    }
+
+    determineType(values) {
+        return extname((values && values[0]) || '').replace('.', '');
+    }
+
+    async publishAssets(tag, entrypoints, options) {
+        Joi.assert(
+            tag,
+            schemas.tag,
+            `Invalid 'tag' argument given when attempting to publish assets.`
+        );
+        Joi.assert(
+            entrypoints,
+            schemas.files,
+            `Invalid 'entrypoints' argument given when attempting to publish assets.`
+        );
+        Joi.assert(
+            options,
+            schemas.options,
+            `Invalid 'options' argument given when attempting to publish assets.`
+        );
+
+        const type = this.determineType(entrypoints);
+        const writer = this.writer(type, entrypoints, options);
+        const { response: { statusCode }, body } = await post({
+            url: url.resolve(
+                this.buildServerUri,
+                `${endpoints.PUBLISH_ASSETS}`
+            ),
+            body: JSON.stringify({
+                tag,
+                type,
+                data: await getStream(writer.bundle()),
+            }),
+            serverId: this.serverId,
+        });
+
+        const { message } = body;
+
+        if (statusCode === 200) {
+            return Promise.resolve(body);
+        }
+
+        if (statusCode === 400) {
+            throw Boom.badRequest(message);
+        }
+
+        throw new Boom(
+            `Asset build server responded with an error. Original message: ${message}.`,
+            {
+                statusCode,
+            }
+        );
+    }
+
+    async publishInstructions(tag, type, data) {
+        Joi.assert(
+            tag,
+            schemas.tag,
+            `Invalid 'tag' argument given when attempting to publish instructions.`
+        );
+        Joi.assert(
+            type,
+            schemas.type,
+            `Invalid 'type' argument given when attempting to publish instructions.`
+        );
+        Joi.assert(
+            data,
+            schemas.bundleInstruction,
+            `Invalid 'data' argument given when attempting to publish instructions.`
+        );
+
+        const { response: { statusCode }, body } = await post({
+            url: url.resolve(
+                this.buildServerUri,
+                `${endpoints.PUBLISH_INSTRUCTIONS}`
+            ),
+            body: JSON.stringify({ tag, type, data }),
+            serverId: this.serverId,
+        });
+
+        const { message } = body;
+
+        if (statusCode === 200) {
+            return Promise.resolve(body);
+        }
+
+        if (statusCode === 400) {
+            throw Boom.badRequest(message);
+        }
+
+        throw new Boom(
```

Comment:
Again, does boom make sense here?

Reply:
Happy to remove. I wouldn't mind if we had better error subclassing going on. Ie. its nice that when we use Joi we get [ValidationError] errors. It would be quite nice to be consistent with our application of error types wherever possible. We could just subclass as we see fit but that might lead to fairly inconsistent error typing across projects. Are there any projects around that can help with this?

Does anyone have any thoughts ideas here? @trygve-lie @SimenB @theneva @sveisvei?

Reply:
This looks interesting: https://www.npmjs.com/package/error

Are we sure that Boom is not appropriate for these particular errors here? Even though this is the client these are http related errors...?

Reply:
One reflection I have is that we already are pretty bad a using existing error types in the language. We tend to just use `Error`.

I'm not sure boom or simmilar libs are suitable for non http errors. 

We could also utilize the System Error which is built into node.js: https://nodejs.org/dist/latest-v9.x/docs/api/errors.html#errors_class_system_error and provide error codes.

Reply:
Yup agree wholeheartedly, would be good to get some useful consistent error types up in things. 

System errors certainly look comprehensive. Perhaps a little low level?